nx/lib/nx.ex

```diff
@@ -15703,6 +15703,87 @@ defmodule Nx do
     sqrt(variance(tensor, opts))
   end
 
+  @doc """
+  Finds the covariance matrix of the input tensor columns.
+
+  Covariance of two random variables X and Y is calculated as mean((X - mean(X)) * (Y - mean(Y))).
+  Covariance matrix element at position (i, j) is equal to
+  the covariance of the i-th and j-th columns of the input tensor.
+  If the `:ddof` (delta degrees of freedom) option is given, the divisor `n - ddof` is used for the sum of the products.
+  Means of the columns can be provided with the `:mean` option. Otherwise, they are estimated using `Nx.mean`.
+
+  ## Examples
+
+    iex> Nx.covariance(Nx.tensor([[1, 2], [3, 4], [5, 6]]))
+    #Nx.Tensor<
+      f32[2][2]
+      [
+        [2.6666667461395264, 2.6666667461395264],
+        [2.6666667461395264, 2.6666667461395264]
+      ]
+    >
+
+    iex> Nx.covariance(Nx.tensor([[1, 2], [3, 4], [5, 6]]), ddof: 1)
+    #Nx.Tensor<
+      f32[2][2]
+      [
+        [4.0, 4.0],
+        [4.0, 4.0]
+      ]
+    >
+
+    iex> Nx.covariance(Nx.tensor([[1, 2], [3, 4], [5, 6]]), mean: Nx.tensor([4, 3]))
+    #Nx.Tensor<
+      f32[2][2]
+      [
+        [3.6666667461395264, 1.6666666269302368],
+        [1.6666666269302368, 3.6666667461395264]
+      ]
+    >
+  """
+  @doc type: :aggregation
+  @spec covariance(tensor :: Nx.Tensor.t(), opts :: Keyword.t()) :: Nx.Tensor.t()
+  def covariance(tensor, opts \\ []) do
+    tensor = to_tensor(tensor)
+    opts = keyword!(opts, ddof: 0, mean: nil)
+
+    if rank(tensor) != 2 do
+      raise ArgumentError, "expected tensor of rank 2, got #{rank(tensor)}"
```

Comment:
We should a least accept tensors of shape `{..., m, n}` where the first k-2 dimensions are batch dimensions.

For the mean, I think we need to receive it as an argument, so we'd have:

```elixir
def covariance(tensor), do: covariance(tensor, Nx.mean(tensor, axes: [-2]), [])
def covariance(tensor, opts) when is_list(opts), do: covariance(tensor, Nx.mean(tensor, axes: [-2]), opts)
def covariance(tensor, mean), do: covariance(tensor, mean, [])
def covariance(tensor, mean, opts) do
...
end
```

Although @josevalim will be able to say for certain if passing the mean as an option is really a problem.

Reply:
And you'd do `mean = to_tensor(mean)` right after `tensor = to_tensor(tensor)`

Reply:
Yes, tensors must be arguments. And I would go with this signature:

```
def covariance(tensor, opts \\ [])
def covariance(tensor, opts) when is_list(opts), do: covariance(tensor, Nx.mean(tensor, axes: [-2]), opts)
def covariance(tensor, mean), do: covariance(tensor, mean, [])
def covariance(tensor, mean, opts) when is_list(opts) do
```

:)

Reply:
I got used to passing arguments as options to the point I forgot Elixir supports multiple-clause functions. 😅

Reply:
> We should at least accept tensors of shape {..., m, n} where the first k-2 dimensions are batch dimensions.

@polvalente Just to clarify: if we take a batch of datasets as input (shape `{..., n, d}`), the output should be a batch of covariances (shape `{..., d, d}`), right?

Reply:
Yes